### PR TITLE
Restructure the time source and clock methods in rcl

### PR DESCRIPTION
**Pull request by tfoote**
This primarily moves the internal references to the TimeSource from inside the Time objects.
It also renames the TimeSource objects to Clocks to follow the work in ros2/design#144

And it lastly remove the default TimeSource logic where now you will need to explicitly connect a TimeSource to a node as we don't want to have global state. 

This is the lowest layer of a larger sweeping change but specific feedback here would be helpful.

### CODE REVIEW

File: rcl/src/rcl/time.c [hunk 1]
```diff
@@ -77,130 +73,127 @@ rcl_get_ros_time(void * data, rcl_time_point_value_t * current_time)
...
+rcl_ros_clock_init(rcl_clock_t * clock)
 {
-  RCL_CHECK_ARGUMENT_FOR_NULL(time_source, RCL_RET_INVALID_ARGUMENT, rcl_get_default_allocator());
-  rcl_init_generic_time_source(time_source);
-  time_source->data = calloc(1, sizeof(rcl_ros_time_source_storage_t));
-  time_source->get_now = rcl_get_ros_time;
-  time_source->type = RCL_ROS_TIME;
+  RCL_CHECK_ARGUMENT_FOR_NULL(clock, RCL_RET_INVALID_ARGUMENT, rcl_get_default_allocator());
+  rcl_init_generic_clock(clock);
+  clock->data = calloc(1, sizeof(rcl_ros_clock_storage_t));
```
Comment (Karsten1987) [inline rcl/src/rcl/time.c]: while refactoring this, this function could also take an allocator to avoid native (c)alloc calls.

**Comment by dirk-thomas** [comment]
@ros2/dev Request for API review. Not ready to be merged yet.

**Comment by tfoote** [comment]
Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3423/)
Linux arm64: [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/671/)
OSX:  [Build Status](http://ci.ros2.org/job/ci_osx/2751/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/3467/)

File: rcl/src/rcl/time.c [hunk 2]
```diff
@@ -77,151 +73,127 @@ rcl_get_ros_time(void * data, rcl_time_point_value_t * current_time)
...
-  RCL_CHECK_ARGUMENT_FOR_NULL(time_source, RCL_RET_INVALID_ARGUMENT, rcl_get_default_allocator());
-  if (time_source->type != RCL_ROS_TIME) {
-    RCL_SET_ERROR_MSG("time_source not of type RCL_ROS_TIME", rcl_get_default_allocator());
+  RCL_CHECK_ARGUMENT_FOR_NULL(clock, RCL_RET_INVALID_ARGUMENT, rcl_get_default_allocator());
+  if (clock->type != RCL_ROS_TIME) {
+    RCL_SET_ERROR_MSG("clock not of type RCL_ROS_TIME", rcl_get_default_allocator());
     return RCL_RET_ERROR;
   }
-  free((rcl_ros_time_source_storage_t *)time_source->data);
+  free((rcl_ros_clock_storage_t *)clock->data);
```
Comment (allenh1) [inline rcl/src/rcl/time.c]: should this be using `rmw_free`?

**Comment by tfoote** [comment]
Updated CI:
Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3440/)
Linux arm64: [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/686/)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/2769/)
WIndows: [Build Status](http://ci.ros2.org/job/ci_windows/3487/)

With three completed it looks like the only failures are the new flake8 changes with a patch proposed in parallel.

**Comment by tfoote** [comment]
Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3452/)
Linux arm64: [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/694/)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/2780/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/3499/)

I went ahead and removed the clock type from the duration. I think it has value, but since chrono doesn't I'll follow their lead.

File: rcl/src/rcl/time.c [hunk 3]
```diff
@@ -23,18 +23,13 @@
...
 #include "rcutils/time.h"
 
-// Process default ROS time sources
-static rcl_time_source_t * rcl_default_ros_time_source;
-static rcl_time_source_t * rcl_default_steady_time_source;
-static rcl_time_source_t * rcl_default_system_time_source;
 
 // Internal storage for RCL_ROS_TIME implementation
-typedef struct rcl_ros_time_source_storage_t
+typedef struct rcl_ros_clock_storage_t
```
Comment (wjwwood) [inline rcl/src/rcl/time.c]: Would `rcl_clock_storage_t` make more sense? Maybe not, not sure.

File: rcl/include/rcl/time.h [hunk 4]
```diff
@@ -45,340 +46,218 @@ typedef rcutils_time_point_value_t rcl_time_point_value_t;
...
   void (* pre_update)(void);
   void (* post_update)(void);
   rcl_ret_t (* get_now)(void * data, rcl_time_point_value_t * now);
   // void (*set_now) (rcl_time_point_value_t);
   void * data;
-} rcl_time_source_t;
+} rcl_clock_t;
 
-struct rcl_ros_time_source_storage_t;
+struct rcl_ros_clock_storage_t;
```
Comment (wjwwood) [inline rcl/include/rcl/time.h]: Does this need to be forward declared? I only see it being used in the `time.c` file.

File: rcl/src/rcl/time.c (hunk 3 above)
Comment (tfoote) [reply rcl/src/rcl/time.c]: This is specific to the ros subtype. It's not used by the system or steady time types.

**Comment by tfoote** [comment]
Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3479/)
Linux arm64: [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/717/)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/2807/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/3524/)

**Comment by tfoote** [comment]
This is now rebased on top of #176 to use the allocator checks.

File: rcl/src/rcl/time.c [hunk 5]
```diff
@@ -77,257 +72,184 @@ rcl_get_ros_time(void * data, rcl_time_point_value_t * current_time)
...
+  RCL_CHECK_ARGUMENT_FOR_NULL(clock, RCL_RET_INVALID_ARGUMENT, rcl_get_default_allocator());
+  if (clock->type != RCL_ROS_TIME) {
     RCL_SET_ERROR_MSG(
       "Time source is not RCL_ROS_TIME cannot enable override.", rcl_get_default_allocator())
     return RCL_RET_ERROR;
   }
-  rcl_ros_time_source_storage_t * storage = \
-    (rcl_ros_time_source_storage_t *)time_source->data;
+  rcl_ros_clock_storage_t * storage = \
+    (rcl_ros_clock_storage_t *)clock->data;
```
Comment (sloretz) [inline rcl/src/rcl/time.c]: This line looks short enough to not need to be wrapped. Same thing on a lot of similar lines below

File: rcl/test/rcl/test_time.cpp [hunk 6]
```diff
@@ -156,25 +159,28 @@ TEST_F(CLASSNAME(TestTimeFixture, RMW_IMPLEMENTATION), test_rcl_ros_time_set_ove
   }
 }
 
-TEST_F(CLASSNAME(TestTimeFixture, RMW_IMPLEMENTATION), test_rcl_init_for_time_source_and_point) {
+TEST_F(CLASSNAME(TestTimeFixture, RMW_IMPLEMENTATION), test_rcl_init_for_clock_and_point) {
   assert_no_realloc_begin();
```
Comment (sloretz) [inline rcl/test/rcl/test_time.cpp]: Does the test need a matching assert_no_realloc_end()?

File: rcl/test/rcl/test_time.cpp [hunk 7]
```diff
@@ -196,101 +202,106 @@ TEST_F(CLASSNAME(TestTimeFixture, RMW_IMPLEMENTATION), test_rcl_init_for_time_so
...
-TEST(CLASSNAME(rcl_time, RMW_IMPLEMENTATION), default_time_source_instanciation) {
-  rcl_time_source_t * ros_time_source = rcl_get_default_ros_time_source();
-  ASSERT_TRUE(rcl_time_source_valid(ros_time_source));
-  rcl_time_source_t * steady_time_source = rcl_get_default_steady_time_source();
-  ASSERT_TRUE(rcl_time_source_valid(steady_time_source));
-  rcl_time_source_t * system_time_source = rcl_get_default_system_time_source();
-  ASSERT_TRUE(rcl_time_source_valid(system_time_source));
+TEST(CLASSNAME(rcl_time, RMW_IMPLEMENTATION), default_clock_instanciation) {
+  rcl_clock_t * ros_clock =
+    reinterpret_cast<rcl_clock_t *>(calloc(1, sizeof(rcl_clock_t)));
```
Comment (sloretz) [inline rcl/test/rcl/test_time.cpp]: Does this need to be `calloc`ed for the test to pass? If someone created `rcl_clock_t` on the stack could random uninitialized values in the structure cause this test to fail?

File: rcl/test/rcl/test_time.cpp (hunk 6 above)
Comment (tfoote) [reply rcl/test/rcl/test_time.cpp]: All the allocation asserts aren't in that test. Removing them all.

File: rcl/test/rcl/test_time.cpp (hunk 7 above)
Comment (tfoote) [reply rcl/test/rcl/test_time.cpp]: It does not need to be calloc specifically.

**Pull request merged**